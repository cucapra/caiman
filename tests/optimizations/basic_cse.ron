// ARGS: -t basic-cse
(
	version : (0, 0, 1),
	program : (
		types : {0:I32},
		funclets : {
			0 : (
                // empty funclet - should remain empty, not cause crashes, etc
                kind : MixedImplicit,
                input_types : [],
                output_types : [],
                nodes : [],
                tail_edge : Return(return_values : []) 
            ),
            1 : (
                // no duplicates, nothing should be changed
                kind : MixedImplicit,
                input_types : [0],
                output_types : [0],
                nodes : [
                    Phi(index : 0),
                    CallExternalCpu(external_function_id : 0, arguments : [0]),
                    ExtractResult(node_id : 1, index : 0),
                ],
                tail_edge : Return(return_values : [2]) 
            ),
            2 : (
                // no duplicates, but with more complex dependencies (multi output)
                kind : MixedImplicit,
                input_types : [0],
                output_types : [0],
                nodes : [
                    Phi(index : 0),
                    CallExternalCpu(external_function_id : 0, arguments : [0]),
                    ExtractResult(node_id : 1, index : 0),
                ],
                tail_edge : Return(return_values : [0, 2]) 
            ),
            3 : (
                // both duplicates dependended on by the same node
                kind : MixedImplicit,
                input_types : [0],
                output_types : [0, 0],
                nodes : [
                    Phi(index : 0),
                    Phi(index : 0),
                    CallExternalCpu(external_function_id : 0, arguments : [0, 1])
                ],
                tail_edge : Return(return_values : [2]) 
            ),
            4 : (
                // both duplicates depended on by the tail
                kind : MixedImplicit,
                input_types : [0],
                output_types : [0, 0],
                nodes : [
                    Phi(index : 0),
                    Phi(index : 0)
                ],
                tail_edge : Return(return_values : [0, 1]) 
            ),
            5 : (
                // more complex: dependency *chains*, duplicates not trivially eq
                kind : MixedImplicit,
                input_types : [0],
                output_types : [0, 0],
                nodes : [
                    Phi(index : 0),
                    Phi(index : 0),
                    CallExternalCpu(external_function_id : 0, arguments : [0]),
                    CallExternalCpu(external_function_id : 0, arguments : [1])
                ],
                tail_edge : Return(return_values : [2, 3]) 
            ),
            6 : (
                // interleaved duplicate chain
                kind : MixedImplicit,
                input_types : [0],
                output_types : [0, 0],
                nodes : [
                    Phi(index : 0),
                    CallExternalCpu(external_function_id : 0, arguments : [0]),
                    Phi(index : 0),
                    CallExternalCpu(external_function_id : 0, arguments : [2])
                ],
                tail_edge : Return(return_values : [1, 3]) 
            ),
            7 : (
                // a sprawling mess of duplication
                kind : MixedImplicit,
                input_types : [0],
                output_types : [0, 0],
                nodes : [
                    Phi(index : 0),
                    CallExternalCpu(external_function_id : 0, arguments : [0]),
                    Phi(index : 0),
                    CallExternalCpu(external_function_id : 0, arguments : [2]),
                    ExtractResult(node_id : 3, index : 0),
                    ExtractResult(node_id : 1, index : 0),
                    CallExternalCpu(external_function_id : 0, arguments : [1, 5]),
                    CallExternalCpu(external_function_id : 0, arguments : [1, 4]),
                    CallExternalCpu(external_function_id : 0, arguments : [6, 7])
                ],
                tail_edge : Return(return_values : [3, 8]) 
            )
        }
	)
)