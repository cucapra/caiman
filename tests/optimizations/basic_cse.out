(
    version: (0, 0, 1),
    program: (
        types: {
            0: I32,
        },
        funclets: {
            0: (
                kind: MixedImplicit,
                input_types: [],
                output_types: [],
                nodes: [],
                tail_edge: Return(
                    return_values: [],
                ),
                input_resource_states: [],
                output_resource_states: [],
                local_meta_variables: {},
            ),
            1: (
                kind: MixedImplicit,
                input_types: [
                    0,
                ],
                output_types: [
                    0,
                ],
                nodes: [
                    Phi(
                        index: 0,
                    ),// [0]
                    CallExternalCpu(
                        external_function_id: 0,
                        arguments: [
                            0,
                        ],
                    ),// [1]
                    ExtractResult(
                        node_id: 1,
                        index: 0,
                    ),
                ],
                tail_edge: Return(
                    return_values: [
                        2,
                    ],
                ),
                input_resource_states: [],
                output_resource_states: [],
                local_meta_variables: {},
            ),
            2: (
                kind: MixedImplicit,
                input_types: [
                    0,
                ],
                output_types: [
                    0,
                ],
                nodes: [
                    Phi(
                        index: 0,
                    ),// [0]
                    CallExternalCpu(
                        external_function_id: 0,
                        arguments: [
                            0,
                        ],
                    ),// [1]
                    ExtractResult(
                        node_id: 1,
                        index: 0,
                    ),
                ],
                tail_edge: Return(
                    return_values: [
                        0,// [0]
                        2,
                    ],
                ),
                input_resource_states: [],
                output_resource_states: [],
                local_meta_variables: {},
            ),
            3: (
                kind: MixedImplicit,
                input_types: [
                    0,
                ],
                output_types: [
                    0,// [0]
                    0,
                ],
                nodes: [
                    Phi(
                        index: 0,
                    ),// [0]
                    CallExternalCpu(
                        external_function_id: 0,
                        arguments: [
                            0,// [0]
                            0,
                        ],
                    ),
                ],
                tail_edge: Return(
                    return_values: [
                        1,
                    ],
                ),
                input_resource_states: [],
                output_resource_states: [],
                local_meta_variables: {},
            ),
            4: (
                kind: MixedImplicit,
                input_types: [
                    0,
                ],
                output_types: [
                    0,// [0]
                    0,
                ],
                nodes: [
                    Phi(
                        index: 0,
                    ),
                ],
                tail_edge: Return(
                    return_values: [
                        0,// [0]
                        0,
                    ],
                ),
                input_resource_states: [],
                output_resource_states: [],
                local_meta_variables: {},
            ),
            5: (
                kind: MixedImplicit,
                input_types: [
                    0,
                ],
                output_types: [
                    0,// [0]
                    0,
                ],
                nodes: [
                    Phi(
                        index: 0,
                    ),// [0]
                    CallExternalCpu(
                        external_function_id: 0,
                        arguments: [
                            0,
                        ],
                    ),
                ],
                tail_edge: Return(
                    return_values: [
                        1,// [0]
                        1,
                    ],
                ),
                input_resource_states: [],
                output_resource_states: [],
                local_meta_variables: {},
            ),
            6: (
                kind: MixedImplicit,
                input_types: [
                    0,
                ],
                output_types: [
                    0,// [0]
                    0,
                ],
                nodes: [
                    Phi(
                        index: 0,
                    ),// [0]
                    CallExternalCpu(
                        external_function_id: 0,
                        arguments: [
                            0,
                        ],
                    ),
                ],
                tail_edge: Return(
                    return_values: [
                        1,// [0]
                        1,
                    ],
                ),
                input_resource_states: [],
                output_resource_states: [],
                local_meta_variables: {},
            ),
            7: (
                kind: MixedImplicit,
                input_types: [
                    0,
                ],
                output_types: [
                    0,// [0]
                    0,
                ],
                nodes: [
                    Phi(
                        index: 0,
                    ),// [0]
                    CallExternalCpu(
                        external_function_id: 0,
                        arguments: [
                            0,
                        ],
                    ),// [1]
                    ExtractResult(
                        node_id: 1,
                        index: 0,
                    ),// [2]
                    CallExternalCpu(
                        external_function_id: 0,
                        arguments: [
                            1,// [0]
                            2,
                        ],
                    ),// [3]
                    CallExternalCpu(
                        external_function_id: 0,
                        arguments: [
                            3,// [0]
                            3,
                        ],
                    ),
                ],
                tail_edge: Return(
                    return_values: [
                        1,// [0]
                        4,
                    ],
                ),
                input_resource_states: [],
                output_resource_states: [],
                local_meta_variables: {},
            ),
        },
        external_cpu_functions: [],
        external_gpu_functions: [],
        value_functions: {},
        pipelines: [],
    ),
)